File: app/src/main/java/com/quran/labs/androidquran/ui/fragment/AyahTranslationFragment.java
package com.quran.labs.androidquran.ui.fragment;

import android.app.Activity;
import android.content.Context;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Button;
import android.widget.ProgressBar;

import com.quran.data.core.QuranInfo;
import com.quran.labs.androidquran.R;
import com.quran.labs.androidquran.common.LocalTranslation;
import com.quran.labs.androidquran.common.QuranAyahInfo;
import com.quran.data.model.VerseRange;
import com.quran.labs.androidquran.presenter.translation.InlineTranslationPresenter;
import com.quran.labs.androidquran.ui.PagerActivity;
import com.quran.labs.androidquran.ui.util.TranslationsSpinnerAdapter;
import com.quran.labs.androidquran.util.QuranSettings;
import com.quran.labs.androidquran.view.InlineTranslationView;
import com.quran.labs.androidquran.view.QuranSpinner;

import java.util.List;
import java.util.Set;

import javax.inject.Inject;

import androidx.annotation.NonNull;

public class AyahTranslationFragment extends AyahActionFragment
    implements InlineTranslationPresenter.TranslationScreen {

  private ProgressBar progressBar;
  private InlineTranslationView translationView;
  private View emptyState;
  private View translationControls;
  private QuranSpinner translator;
  private TranslationsSpinnerAdapter translationAdapter;
  private List<LocalTranslation> translations;

  @Inject QuranInfo quranInfo;
  @Inject QuranSettings quranSettings;
  @Inject InlineTranslationPresenter translationPresenter;

  @Override
  public void onAttach(@NonNull Context context) {
    super.onAttach(context);
    ((PagerActivity) getActivity()).getPagerActivityComponent().inject(this);
  }

  @Override
  public View onCreateView(LayoutInflater inflater,
                           ViewGroup container, Bundle savedInstanceState) {
    final View view = inflater.inflate(
        R.layout.translation_panel, container, false);

    translator = view.findViewById(R.id.translator);
    translationView = view.findViewById(R.id.translation_view);
    progressBar = view.findViewById(R.id.progress);
    emptyState = view.findViewById(R.id.empty_state);
    translationControls = view.findViewById(R.id.controls);
    final View next = translationControls.findViewById(R.id.next_ayah);
    next.setOnClickListener(onClickListener);

    final View prev = translationControls.findViewById(R.id.previous_ayah);
    prev.setOnClickListener(onClickListener);

    final Button getTranslations =
        view.findViewById(R.id.get_translations_button);
    getTranslations.setOnClickListener(onClickListener);
    return view;
  }

  @Override
  public void onResume() {
    // currently needs to be before we call super.onResume
    translationPresenter.bind(this);
    super.onResume();
  }

  @Override
  public void onPause() {
    translationPresenter.unbind(this);
    super.onPause();
  }

  private final View.OnClickListener onClickListener = v -> {
    final Activity activity = getActivity();
    if (activity instanceof PagerActivity) {
      final PagerActivity pagerActivity = (PagerActivity) activity;

      switch (v.getId()) {
        case R.id.get_translations_button:
          pagerActivity.startTranslationManager();
          break;
        case R.id.next_ayah:
          pagerActivity.nextAyah();
          break;
        case R.id.previous_ayah:
          pagerActivity.previousAyah();
          break;
      }
    }
  };

  @Override
  public void refreshView() {
    if (start == null || end == null) { return; }

    final Activity activity = getActivity();
    if (activity instanceof PagerActivity) {
      PagerActivity pagerActivity = (PagerActivity) activity;
      if (translations == null || translations.size() == 0) {
        translations = pagerActivity.getTranslations();
      }

      if (translations == null || translations.size() == 0) {
        progressBar.setVisibility(View.GONE);
        emptyState.setVisibility(View.VISIBLE);
        translationControls.setVisibility(View.GONE);
        return;
      }

      if (translationAdapter == null) {
        Set<String> activeTranslations = pagerActivity.getActiveTranslations();
        if (activeTranslations == null) {
          activeTranslations = quranSettings.getActiveTranslations();
        }

        translationAdapter = new TranslationsSpinnerAdapter(activity,
            R.layout.translation_ab_spinner_item,
            pagerActivity.getTranslationNames(),
            translations,
            activeTranslations,
            selectedItems -> {
              quranSettings.setActiveTranslations(selectedItems);
              refreshView();
            });
        translator.setAdapter(translationAdapter);
      }

      if (start.equals(end)) {
        translationControls.setVisibility(View.VISIBLE);
      } else {
        translationControls.setVisibility(View.GONE);
      }

      final int verses = 1 + Math.abs(
          quranInfo.getAyahId(start.sura, start.ayah) - quranInfo.getAyahId(end.sura, end.ayah));
      VerseRange verseRange = new VerseRange(start.sura, start.ayah, end.sura, end.ayah, verses);
      translationPresenter.refresh(verseRange);
    }
  }

  @Override
  public void setVerses(@NonNull LocalTranslation[] translations, @NonNull List<QuranAyahInfo> verses) {
    progressBar.setVisibility(View.GONE);
    if (verses.size() > 0) {
      emptyState.setVisibility(View.GONE);
      translationView.setAyahs(translations, verses);
    } else {
      emptyState.setVisibility(View.VISIBLE);
    }
  }
}

File: app/src/main/java/com/quran/labs/androidquran/ui/fragment/AyahTranslationFragment.kt
package com.quran.labs.androidquran.ui.fragment

import android.app.Activity
import android.content.Context
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.Button
import android.widget.ProgressBar
import com.quran.data.core.QuranInfo
import com.quran.data.model.VerseRange
import com.quran.labs.androidquran.R
import com.quran.labs.androidquran.common.QuranAyahInfo
import com.quran.labs.androidquran.presenter.translation.InlineTranslationPresenter
import com.quran.labs.androidquran.presenter.translation.InlineTranslationPresenter.TranslationScreen
import com.quran.labs.androidquran.presenter.translationlist.TranslationListPresenter
import com.quran.labs.androidquran.ui.PagerActivity
import com.quran.labs.androidquran.ui.helpers.SlidingPagerAdapter
import com.quran.labs.androidquran.ui.util.TranslationsSpinnerAdapter
import com.quran.labs.androidquran.util.QuranSettings
import com.quran.labs.androidquran.view.InlineTranslationView
import com.quran.labs.androidquran.view.QuranSpinner
import com.quran.mobile.di.AyahActionFragmentProvider
import com.quran.mobile.translation.model.LocalTranslation
import kotlinx.coroutines.MainScope
import kotlinx.coroutines.cancel
import kotlinx.coroutines.flow.launchIn
import kotlinx.coroutines.flow.onEach
import kotlinx.coroutines.launch
import javax.inject.Inject
import kotlin.math.abs

class AyahTranslationFragment : AyahActionFragment(), TranslationScreen {
  private lateinit var progressBar: ProgressBar
  private lateinit var translationView: InlineTranslationView
  private lateinit var emptyState: View
  private lateinit var translationControls: View
  private lateinit var translator: QuranSpinner

  private var translationAdapter: TranslationsSpinnerAdapter? = null

  @Inject
  lateinit var quranInfo: QuranInfo

  @Inject
  lateinit var quranSettings: QuranSettings

  @Inject
  lateinit var translationPresenter: InlineTranslationPresenter

  private val scope = MainScope()

  object Provider : AyahActionFragmentProvider {
    override val order = SlidingPagerAdapter.TRANSLATION_PAGE
    override val iconResId = com.quran.labs.androidquran.common.toolbar.R.drawable.ic_translation
    override fun newAyahActionFragment() = AyahTranslationFragment()
  }

  override fun onAttach(context: Context) {
    super.onAttach(context)
    (activity as? PagerActivity)?.pagerActivityComponent?.inject(this)
  }

  override fun onDetach() {
    scope.cancel()
    super.onDetach()
  }

  override fun onCreateView(
    inflater: LayoutInflater,
    container: ViewGroup?,
    savedInstanceState: Bundle?
  ): View? {
    val view = inflater.inflate(
      R.layout.translation_panel, container, false
    )
    translator = view.findViewById(R.id.translator)
    translationView = view.findViewById(R.id.translation_view)
    progressBar = view.findViewById(R.id.progress)
    emptyState = view.findViewById(R.id.empty_state)
    translationControls = view.findViewById(R.id.controls)

    val next = translationControls.findViewById<View>(R.id.next_ayah)
    next.setOnClickListener(onClickListener)
    val prev = translationControls.findViewById<View>(R.id.previous_ayah)
    prev.setOnClickListener(onClickListener)
    val getTranslations = view.findViewById<Button>(R.id.get_translations_button)
    getTranslations.setOnClickListener(onClickListener)
    return view
  }

  override fun onResume() {
    // currently needs to be before we call super.onResume
    translationPresenter.bind(this)
    super.onResume()
  }

  override fun onPause() {
    translationPresenter.unbind(this)
    super.onPause()
  }

  private val onClickListener = View.OnClickListener { v: View ->
    val activity: Activity? = activity
    if (activity is PagerActivity) {
      when (v.id) {
        R.id.get_translations_button -> activity.startTranslationManager()
        R.id.next_ayah -> readingEventPresenter.selectNextAyah()
        R.id.previous_ayah -> readingEventPresenter.selectPreviousAyah()
      }
    }
  }

  override fun onTranslationsUpdated(translations: List<LocalTranslation>) {
    if (translations.isEmpty()) {
      progressBar.visibility = View.GONE
      emptyState.visibility = View.VISIBLE
      translationControls.visibility = View.GONE
      translator.visibility = View.GONE
      translationView.visibility = View.GONE
    } else {
      val activeTranslationsFilesNames = quranSettings.activeTranslations

      val adapter = translationAdapter
      if (adapter == null) {
        translationAdapter = TranslationsSpinnerAdapter(
          activity,
          R.layout.translation_ab_spinner_item,
          translations.map { it.resolveTranslatorName() }.toTypedArray(),
          translations,
          activeTranslationsFilesNames,
        ) { selectedItems: Set<String?>? ->
          quranSettings.activeTranslations = selectedItems
          // this is the refresh for when a translation is selected from the spinner
          refreshView()
        }
        translator.adapter = translationAdapter
      } else {
        adapter.updateItems(
          translations.map { it.resolveTranslatorName() }.toTypedArray(),
          translations,
          activeTranslationsFilesNames
        )
      }
      refreshView()
    }
  }

  public override fun refreshView() {
    val start = start
    val end = end
    if (start == null || end == null) {
      return
    }

    val verses = 1 + abs(
      quranInfo.getAyahId(start.sura, start.ayah) - quranInfo.getAyahId(end.sura, end.ayah)
    )
    val verseRange = VerseRange(start.sura, start.ayah, end.sura, end.ayah, verses)
    scope.launch {
      translationPresenter.refresh(verseRange)
    }
  }

  override fun setVerses(translations: Array<LocalTranslation>, verses: List<QuranAyahInfo>) {
    progressBar.visibility = View.GONE
    if (verses.isNotEmpty()) {
      emptyState.visibility = View.GONE
      translationControls.visibility = View.VISIBLE
      translator.visibility = View.VISIBLE
      translationView.visibility = View.VISIBLE
      translationView.setAyahs(translations, verses)
    } else {
      emptyState.visibility = View.VISIBLE
    }
  }
}
